File: go/01-que-es-un-agente/clasificador-nivel.go
// Test de localizacion: clasifica un sistema en el espectro smolagents.

// Cómo ejecutar: make go FILE=go/01-que-es-un-agente/clasificador-nivel.go

package main

import "fmt"

type features struct {
	multiAgente               bool
	codeAgent                 bool
	loopNoAcotadoYDecideTools bool
	loopAcotadoConTools       bool
	llmEligeRutaSinLoop       bool
}

func classify(f features) string {
	switch {
	case f.multiAgente:
		return "★★★ multi-agente"
	case f.codeAgent:
		return "★★★ code agent"
	case f.loopNoAcotadoYDecideTools:
		return "★★☆ multi-step"
	case f.loopAcotadoConTools:
		return "★★☆ tool caller"
	case f.llmEligeRutaSinLoop:
		return "★☆☆ router"
	default:
		return "☆☆☆ procesador"
	}
}

func main() {
	cases := []struct {
		name string
		f    features
	}{
		{"traduccion sin tools", features{}},
		{"router por intent", features{llmEligeRutaSinLoop: true}},
		{"RAG simple (1 retrieve + 1 generate)", features{loopAcotadoConTools: true}},
		{"ReAct hasta end_turn", features{loopNoAcotadoYDecideTools: true}},
		{"supervisor + workers", features{multiAgente: true}},
		{"agente que escribe codigo Python", features{codeAgent: true}},
	}
	for _, c := range cases {
		fmt.Printf("%-25s  %s\n", classify(c.f), c.name)
	}
}

File: typescript/01-que-es-un-agente/clasificador-nivel.ts
// Test de localizacion: clasifica un sistema en el espectro smolagents.

// Cómo ejecutar: make ts SCRIPT=typescript/01-que-es-un-agente/clasificador-nivel.ts
// Qué esperar: 6 casos clasificados en el espectro de autonomia. Sin llamadas API.

type Features = {
  multiAgente?: boolean;
  codeAgent?: boolean;
  loopNoAcotadoYDecideTools?: boolean;
  loopAcotadoConTools?: boolean;
  llmEligeRutaSinLoop?: boolean;
};

function classify(f: Features): string {
  if (f.multiAgente) return "★★★ multi-agente";
  if (f.codeAgent) return "★★★ code agent";
  if (f.loopNoAcotadoYDecideTools) return "★★☆ multi-step";
  if (f.loopAcotadoConTools) return "★★☆ tool caller";
  if (f.llmEligeRutaSinLoop) return "★☆☆ router";
  return "☆☆☆ procesador";
}

const CASES: Array<[string, Features]> = [
  ["traduccion sin tools", {}],
  ["router por intent", { llmEligeRutaSinLoop: true }],
  ["RAG simple (1 retrieve + 1 generate)", { loopAcotadoConTools: true }],
  ["ReAct hasta end_turn", { loopNoAcotadoYDecideTools: true }],
  ["supervisor + workers", { multiAgente: true }],
  ["agente que escribe codigo Python", { codeAgent: true }],
];

for (const [name, f] of CASES) {
  console.log(classify(f).padEnd(25) + "  " + name);
}
